src/verify.rs

```diff
@@ -89,15 +89,35 @@ use self::PreparedStep::*;
 ///
 /// This type would be Copy except for the fact that the bitset can require
 /// overflow storage :(.
-struct StackSlot {
+#[derive(Clone)]
+pub struct StackSlot {
     vars: Bitset,
     code: Atom,
     expr: Range<usize>,
 }
 
+/// A constructor trait for plugging in to the verifier, to collect extra data during the
+/// verification pass
+pub trait ProofBuilder {
```

Comment:
I was going to write a separate proof output pass in `proof.rs` but there is a *huge* amount of overlap. I'll see if I can type-generalize the `Vec<()>` structure into `()` as well.